airflow/contrib/hooks/google_discovery_api_hook.py

```diff
@@ -0,0 +1,163 @@
+    :param api_version: The version of the api that will be requested for example 'v3'.
+    :type api_version: str
+    :param gcp_conn_id: The connection ID to use when fetching connection info.
+    :type gcp_conn_id: str
+    :param delegate_to: The account to impersonate, if any.
+        For this to work, the service account making the request must have
+        domain-wide delegation enabled.
+    :type delegate_to: str
+    """
+    _conn = None
+
+    def __init__(
+        self, api_service_name, api_version, gcp_conn_id='google_cloud_default', delegate_to=None
+    ):
+        super(GoogleDiscoveryApiHook, self).__init__(gcp_conn_id=gcp_conn_id, delegate_to=delegate_to)
+        self.api_service_name = api_service_name
+        self.api_version = api_version
+
+    def get_conn(self):
+        """
+        Creates an authenticated api client for the given api service name and credentials.
+
+        :return: the authenticated api service.
+        :rtype: Resource
+        """
+        self.log.info("Authenticating Google API Client")
+
+        if not self._conn:
+            http_authorized = self._authorize()
+            self._conn = build(serviceName=self.api_service_name, version=self.api_version,
+                               http=http_authorized, cache_discovery=False)
+        return self._conn
+
+    def query(self, endpoint, data, paginate=False):
+        """
+        Creates a dynamic API call to any Google API registered in Google's API Client Library
+        and queries it.
+
+        :param endpoint: The client libraries path to the api call's executing method.
+            For example: 'analyticsreporting.reports.batchGet'
+
+            .. seealso:: https://developers.google.com/apis-explorer
+                for more information on what methods are available.
+        :type endpoint: str
+        :param data: The data (endpoint params) needed for the specific request to given endpoint.
+        :type data: dict
+        :param paginate: If set to True, it will collect all pa
+        :type paginate: bool
+        :return: the API response from the passed endpoint.
+        :rtype: dict
+        """
+        google_api_conn_client = self.get_conn()
+
+        api_response = self._call_api_request(
+            google_api_conn_client, endpoint, data, paginate
+        )
+        return api_response
+
+    def _call_api_request(
+        self, google_api_conn_client, endpoint, data, paginate
+    ):
+        api_endpoint_parts = endpoint.split('.')
+
+        google_api_endpoint_instance = self._build_api_request(
+            google_api_conn_client,
+            api_sub_functions=api_endpoint_parts[1:],
+            api_endpoint_params=data
+        )
+
+        if paginate:
+            return self._paginate_api(
+                google_api_endpoint_instance, google_api_conn_client,
+                api_endpoint_parts
+            )
+
+        return google_api_endpoint_instance.execute()
```

Comment:
I think it would be a good idea to use `num_retries` in execute:
```suggestion
        return google_api_endpoint_instance.execute(num_retries=self.num_retries)
```

Reply:
Very good point! - will add this.

Reply:
I added it as a local argument to the `query` function. WDYT @nuclearpinguin ?

Reply:
I am ok with this approach ✅